### ISSUE TITLE
When using OpenID for SSO, the `/auth/login/<provider>` URI is cached by the browser

### ISSUE THREAD

**Issue by g-pichler** [directus/directus#23999]
### Describe the Bug

When using OpenID for SSO, the `/auth/login/<provider>?redirect=...` URI results in a `302 Redirect`. However, it sets at least one important cookie, `openid.<provider>` which is essential for the login flow. Thus, this request should not be cached. This is not communicated to the browser in the headers, as the `cache-control` header is not set in the reply.
In some cases this will lead to `INVALID_CREDENTIALS` errors.

In particular, when restoring a webapp, after restarting the browser, firefox skips opening `/auth/login/<provider>?redirect=...` and directly issues a `GET` request to the auth provider. This fails as the required cookie is missing.

We would propose to add `Header: cache-control: no-store, must-revalidate, max-age=0` when responding to the `GET/ auth/login/<provider>?redirect=...` request.

### To Reproduce

This caching issue only seems to affect the login flow only in narrow circumstance. To us, it is nevertheless important:
- Start Firefox and configure it to _"Open previous windows and tabs"_ (in _Settings_ → _General_ → _Startup_)
- Use OpenID SSO to login to a web app by navigating to `directus.example.com/auth/login/example_provider?redirect=https://webapp.example.com/`.
- With the open tab `https://webapp.example.com/`, close firefox.
- Re-Open firefox
- You see a webpage with a JSON error
  ```
  {"errors":[{"message":"Invalid user credentials.","extensions":{"code":"INVALID_CREDENTIALS"}}]}
  ```

### Directus Version

v11.1.0

### Hosting Strategy

Self-Hosted (Custom)

### Database

PostgreSQL 14

**Comment by squallttthp** [comment]
I have the same error as you. I found the problem in the environment variable configuration below:

Use session mode instead of cookie mode:
AUTH_GOOGLE_MODE="session"
SESSION_COOKIE_DOMAIN="directus.myserver.com" # your server domain
SESSION_COOKIE_SECURE="true"
SESSION_COOKIE_SAME_SITE="None"

AUTH_PROVIDERS="google"
# url callback in client domain
AUTH_GOOGLE_REDIRECT_ALLOW_LIST="http://localhost:3000/login/google/callback"

I hope you're able to resolve the problem

**Comment by g-pichler** [comment]
Thanks for the hint. We did try `session` mode, but found that the directus app was only compatible with `cookie`.

> I have the same error as you. I found the problem in the environment variable configuration below:
> 
> Use session mode instead of cookie mode: AUTH_GOOGLE_MODE="session" SESSION_COOKIE_DOMAIN="directus.myserver.com" # your server domain SESSION_COOKIE_SECURE="true" SESSION_COOKIE_SAME_SITE="None"
> 
> AUTH_PROVIDERS="google" AUTH_GOOGLE_REDIRECT_ALLOW_LIST="http://localhost:3000/login/google/callback" # url callback in client domain
> 
> I hope you're able to resolve the problem

**Comment by g-pichler** [comment]
> Good to know: Google SSO works on the directus app (backend), but the redirect to localhost does not include any parameters or JSON. So I#m not able to set a cookie, login, etc.
> 
> Any idea what settings to check?

Maybe I misundertand.

But, I think this might be by design. The redirect to `localhost` does not include any information, as `directus` is logged in to the Google SSO, not your webapp. You can use an `XHR` request in your webapp to directus e.g. `https://my.directus.com/users/me`, to obtain the user credentials or to any other `https://my.directus.com/...` URL to perform actions as the the authenticated user.

**Comment by chrislsww** [comment]
@g-pichlert thanks for your quick reply. To clarify, these are different tests.

1. The `localhost` test was made in a different anonymous browser, simulating the user login flow (no cookies set). In my user example, the user will never visit the directus app and should stay on the client all the time.

2. To test the oauth flow with Google and Directus in general, I (as admin) tried to login through my `directus app` at `/admin/login`. This was just to verify the setup with Google works. 

So the issue remains: 

1. Login from client (frontend) - works ✅ 
2. redirect to Google for authentication - works ✅ 
3. Directus receiving `?code=` etc from google verifying the user - works ✅ 
4. Directus creating `access_token` and `refresh_token` - works ✅ 
5. Diretus redirecting to my client including the token - **not working!** ❌ 

The redirect to `http://localhost:5173/auth/callback` generally works, but there are **no parameters in the URL, nor is there any JSON payload to work with**.

**Comment by g-pichler** [comment]
I understand. I believe I understood correctly and still think this is by design. Maybe a directus dev or somebody else can explain this better.

From my understanding, the callback to `http://localhost:5173/auth/callback` is not supposed to get parameters from SSO. Even though your client never visits the directus **app**, they still communicate with directus in the background, through the REST/graphql api. Only directus is authenticated to SSO, not your webapp. If you want your webapp to authenticate, you have to go through SSO authentication with your provider (Google?) independently from directus.

In any case, I do not think this bug report is the correct place for your issue, nor related to it.

**Comment by chrislsww** [comment]
You are right @g-pichler , this is by design and no bug. I got a bit confuse by the many discussions and approaches in different channels and different approaches to use SSO (with and without Directus).

In the [Directus documenatation](https://docs.directus.io/self-hosted/sso.html#seamless-sso), is says:

> On your login page, following the example of https://client.myserver.com/login, you need to call the refresh endpoint either via REST API or via SDK in order to have a session cookie or an access_token. 

REST API example:
```
await fetch('https://directus.myserver.com/auth/refresh', {
	method: 'POST',
	credentials: 'include', // this is required in order to send the refresh/session token cookie
	headers: { 'Accept': 'application/json', 'Content-Type': 'application/json' },
	body: JSON.stringify({ mode: 'session' }) // using 'session' mode, but can also be 'cookie' or 'json'
});
```
